crates/ironrdp-server/src/server.rs

```diff
@@ -0,0 +1,249 @@
+};
+
+/// RDP Server
+///
+/// A server is created to listen for connections.
+/// After the connection sequence is finalized using the provided security mechanism, the server can:
+///  - receive display updates from a [`RdpServerDisplay`] and forward them to the client
+///  - receive input events from a client and forward them to an [`RdpServerInputHandler`]
+///
+/// # Example
+///
+/// ```
+/// let mut server = RdpServer::builder()
+///     .with_addr(([127, 0, 0, 1], 3389))
+///     .with_ssl(tls_acceptor)
+///     .with_input_handler(handler)
+///     .with_display_handler(display)
+///     .build();
+///
+/// server.run().await
+/// ```
+pub struct RdpServer {
+    opts: RdpServerOptions,
+    handler: Box<dyn RdpServerInputHandler>,
+    display: Box<dyn RdpServerDisplay>,
+}
+
+impl RdpServer {
+    pub fn new<H, D>(opts: RdpServerOptions, handler: H, display: D) -> Self
+    where
+        H: RdpServerInputHandler + 'static,
+        D: RdpServerDisplay + 'static,
+    {
+        Self {
+            opts,
+            handler: Box::new(handler),
+            display: Box::new(display),
+        }
+    }
+
+    pub fn builder<H, D>() -> builder::RdpServerBuilder<builder::WantsAddr, H, D>
+    where
+        H: RdpServerInputHandler,
+        D: RdpServerDisplay,
+    {
+        builder::RdpServerBuilder::new()
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        let listener = TcpListener::bind(self.opts.addr).await?;
+
+        while let Ok((stream, peer)) = listener.accept().await {
+            debug!("received connection from {:?}", peer);
+
+            let size = self.display.size().await;
+            let capabilities = capabilities::capabilities(&self.opts, size.clone());
+            let mut acceptor = ServerAcceptor::new(self.opts.clone(), size, capabilities);
+
+            match acceptor::accept_begin(stream, &mut acceptor).await {
+                Ok(BeginResult::ShouldUpgrade(stream)) => {
+                    let upgraded = acceptor::upgrade(&self.opts.security, stream).await?;
+                    let framed = TokioFramed::new(upgraded);
+                    let (framed, _) = acceptor::accept_finalize(framed, &mut acceptor).await?;
+                    self.client_loop(framed).await?;
+                }
+
+                Ok(BeginResult::Continue(framed)) => {
+                    let (framed, _) = acceptor::accept_finalize(framed, &mut acceptor).await?;
+                    self.client_loop(framed).await?;
+                }
+
+                Err(e) => {
+                    eprintln!("connection error: {:?}", e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn client_loop<S>(&mut self, mut framed: Framed<S>) -> Result<()>
+    where
+        S: FramedWrite + FramedRead,
+    {
+        let mut buffer = vec![0u8; 4096];
+        let mut encoder = UpdateEncoder::new();
+
+        debug!("starting client loop");
+
+        'main: loop {
+            select! {
```

Comment:
`select!` seems wrong here, as it will terminate one of the branch: "Polls multiple futures and streams simultaneously, executing the branch for the future that finishes first". So you may have unexpected errors when looping and continue reading the input stream for ex, or unfinished display updates. It should spawn tasks instead.

Reply:
A similar loop is used in the `ironrdp-client`: https://github.com/mihneabuz/IronRDP/blob/master/crates/ironrdp-client/src/rdp.rs#L135C19-L135C19

From what I understand, dropping either of the futures should have no effect if they are not yet completed.
Polling `read_pdu` will read into an internal buffer, and only return the frame when enough bytes are read. Thus, it should pick up from where it left off on the next iteration of the loop.
The display should also act as a channel/receiver, so dropping it early should not lose any items. And once an update is received, it will block the loop until it is sent completely.

Spawning tasks might be better for performance here, so we don't block the input events while we send display updates. Or just improve the display update encoding. I was hoping to do this in a follow-up pr :)

Reply:
`read_pdu` might be ok, but the display update write loop doesn't look ok to me.

Reply:
Are you referring to the `self.display.get_update()` below that is used as a select branch event? Indeed, depending on the implementation, this trait method may not be cancellation safe. Documentation of `get_update` should specify that the implementer must ensure the returned future is cancellation safe. In case you are referring to the `framed.write_all(&buffer[..len]).await` in the branch itself, this is not a problem: cancellation safety is required only for the futures used as "events" of the `select!`. The branch itself is never cancelled once it is entered. The only concern here is that we will stop reading PDUs until we are done writing the update, I’m okay with waiting for a follow up PR (though this may be addressed immediately as well).